infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -126,85 +101,127 @@ public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull R
+        }
 
-            conflictCache.invalidate(key);
-            entries.incrementAndGet();
-        });
+        int i = 0;
+        while (!cacheEntries.isEmpty() && cacheEntries.peek().txVersion == firstEntry.txVersion) {
+            ConflictTxStream entry = cacheEntries.poll();
+            cacheConflictKeys.remove(entry);
+            cacheEntriesBytes -= entry.size();
+            i++;
+        }
 
-        log.info("Invalidated entries: {}", entries.get());
+        log.trace("Evict {} entries", i);
+        maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
+        return i;
     }
 
     /**
-     * The cache size
+     * Invalidate all records up to a trim mark (not included).
      *
-     * @return cache size
+     * @param trimMark trim mark
      */
-    public long size() {
-        return conflictCache.estimatedSize();
+    public void invalidateUpTo(long trimMark) {
+        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
+        int entries = 0;
+        int pqEntries = 0;
+        while(Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
+            pqEntries +=invalidateFirst();
+            entries++;
+        }
+        log.info("Invalidated entries {} addresses {}", pqEntries, entries);
     }
 
     /**
-     * Put a value in the cache
+     * The cache size as the number of entries
      *
-     * @param conflictStream conflict stream
-     * @param newTail        global tail
+     * @return cache size
      */
-    public void put(ConflictTxStream conflictStream, long newTail) {
-        conflictCache.put(conflictStream, newTail);
+    public int size() {
+        return cacheConflictKeys.size();
     }
 
     /**
-     * Discard all entries in the cache
+     * The memory space used by the entries and also the space used by
+     * priorityque, hashmap to store the pointers
+     * @return the memory space used in bytes:
      */
-    public void invalidateAll() {
-        log.info("Invalidate sequencer server cache");
-        conflictCache.invalidateAll();
+    public long byteSize() {
+        return cacheEntriesBytes + this.size()*ENTRY_OVERHEAD;
     }
 
-    /**
-     * Update max conflict wildcard by a new address
+    /*
+     * Put a value in the cache
      *
-     * @param newMaxConflictWildcard new conflict wildcard
+     * @param conflictStream conflict stream
      */
-    public void updateMaxConflictAddress(long newMaxConflictWildcard) {
-        log.info("updateMaxConflictAddress, new address: {}", newMaxConflictWildcard);
-        maxConflictWildcard = newMaxConflictWildcard;
-        maxConflictNewSequencer = newMaxConflictWildcard;
+    public boolean put(ConflictTxStream conflictStream) {
+
+        Long val = cacheConflictKeys.get(conflictStream);
+        if (val != null && val >= conflictStream.txVersion) {
+            log.error("For key {}, the value {} is not smaller  than the expected value {} or maxWildCard",
```

Comment:
Please remove extra space after "smaller"
Also the log says "or maxWildCard" but the code does not check maxWildCard.

Reply:
Done.